docs/rfcs/034-storage-controller-restarts.md

```diff
@@ -0,0 +1,208 @@
+to issue reconciles.
+* Have one uniform implementation for restarts and upgrades
+* Fit in with the current Kubernetes deployment scheme
+
+## Non Goals
+
+* Implement our own consensus algorithm from scratch
+* Completely eliminate downtime storage controller downtime. Instead we aim to reduce it to the point where it looks
+like a transient error to the control plane
+
+## Impacted Components
+
+* storage controller
+* deployment orchestration (i.e. Ansible)
+* helm charts
+
+## Terminology
+
+* Observed State: in-memory mapping between tenant shards and their current pageserver locations - currently built up
+at start-up by quering pageservers
+* Deployment: Kubernetes [primitive](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) that models
+a set of replicas
+
+## Implementation
+
+### High Level Flow
+
+At a very high level the proposed idea is to start a new storage controller instance while
+the previous one is still running and cut-over to it when it becomes ready. The new instance,
+should coordinate with the existing one and transition responsibility gracefully. While the controller
+has built in safety against split-brain situations (via generation numbers), we'd like to avoid such
+scenarios since they can lead to availability issues for tenants that underwent changes while two controllers
+were operating at the same time and require operator intervention to remedy.
+
+### Kubernetes Deployment Configuration
+
+On the Kubernetes configuration side, the proposal is to update the storage controller `Deployment`
+to use `spec.strategy.type = RollingUpdate`, `spec.strategy.rollingUpdate.maxSurge=1` and `spec.strategy.maxUnavailable=0`.
+Under the hood, Kubernetes creates a new replica set and adds one pod to it (`maxSurge=1`). The old replica set does not
+scale down until the new replica set has one replica in the ready state (`maxUnavailable=0`).
+
+The various possible failure scenarios are investigated in the [Handling Failures](#handling-failures) section.
+
+### Storage Controller Start-Up
+
+This section describes the primitives required on the storage controller side and the flow of the happy path.
+
+#### Database Table For Leader Synchronization
+
+A new table should be added to the storage controller database for leader synchronization during startup.
+This table will always contain at most one row. The proposed name for the table is `leader` and the schema
+contains two elements:
+* `hostname`: represents the hostname for the current storage controller leader - should be addressible
+from other pods in the deployment
+* `start_timestamp`: holds the start timestamp for the current storage controller leader (UTC timezone) - only required
+for failure case handling: see [Previous Leader Crashes Before New Leader Readiness](#previous-leader-crashes-before-new-leader-readiness)
+
+Storage controllers will read the leader row at start-up and then update it to mark themselves as the leader
```

Comment:
To avoid multiple controllers potentially doing their start-up dances concurrently, what if (in the happy path where we get a step_down response) we require a controller to gain leadership in the database before proceeding to start up?

So it would be:
1. Call step down, get observed state
2. Claim leadership in the database
3. Apply observed state, make ourselves available.

I think the current proposal is like that but with 3 before 2.

Reply:
I guess we could, but I don't like the idea of having a separate code path and behaviour for this case. It's easier to reason about when all cases are treated the same way.

Concurrent start-up dances should be safe by design via the compare-exchange semantics proposed and there won't be a whole lot left in the startup sequence either:
1. Call step-down
2. Load state from database
3. Load observed state into memory
4. Heartbeat nodes (could be moved later - not sure yet)
5. Mark self as leader
6. Reschedule and reconcile everything

Reply:
I've hopefully clarified the proposed start-up sequence in https://github.com/neondatabase/neon/pull/8310/commits/503692fbef6ce05bf000aae864c53fbf76d6588f